### DISCUSSION TITLE
Wagtail energy transparency and energy efficiency

### DISCUSSION THREAD

**Discussion by ArneTR** [wagtail/wagtail#10535] [Google Summer of Code]
Hey all,

this posting originated as an idea after a call with @mrchrisadams @laymonage @thibaudcolas @PaarthAgarwal @ribalba and @salty-ivy  to start an energy related project for Wagtail.

I work at Green Coding Berlin (GCB) in Germany where we develop open source software for energy and carbon measurement of software and will contribute during the GSoC as a mentor. In this post I want to provide some overview how we at GCB typically approach energy releated topics in Software and spark some ideas for bits and pieces to work on during the GSoC.

1. Overview
For people totally new to the topic I can really recommend the [great article](https://eco.kde.org/handbook/) by [Joseph P. De Veaugh-Geiss](https://invent.kde.org/joseph) about general emissions of software, it's current role in the whole climate topic and some patterns how a German Eco Label (Blue Angel for Software) tries to approach the topic.

Generally the notion is software uses energy and to be more efficient this usage has to be brought down. The most logical step is thus to first get a measurement and then try to improve it.

## 1. Obtaining energy related metrics

For energy metrics some specialized tools exist that typically spit out a value in Joules, Watts (which is effectively Joules per second) or kilo-Watt-hours (kWh).

The [Green Software Foundation](https://greensoftware.foundation/) provides [a good overview list](https://github.com/Green-Software-Foundation/awesome-green-software).

To be highlighted as starter / prototyping tools I guess shall be:
- [Scaphandre](https://github.com/hubblo-org/scaphandre)
- [perf_events](https://stackoverflow.com/questions/55956287/perf-power-consumption-measure-how-does-it-work)
- [Intel onboard tools for Windows](https://www.intel.com/content/www/us/en/developer/articles/tool/power-gadget.html)
- [macOS onboard tools](https://www.green-coding.berlin/blog/power-measurement-on-macos/)
- [CodeCarbon.io](https://codecarbon.io/) Python library to measure energy
![Screenshot 2023-06-10 at 9 19 26 AM](https://github.com/wagtail/wagtail/assets/250671/2d18a288-f2b9-4ad9-81b6-3e96dadee554)
Scaphandre demo picture

These tools all use functionalities that come along on modern Intel, AMD or M1 processors and can give the energy of many main components like CPU, RAM, onboard-graphics, chipsets etc.

The benefit is that this allows for capturing the most relevant metrics so that energy is captured accurately enough to make optimizations and one does not need to buy a power meter.

Stuff like networking, hard-disks, monitor, embodied carbon etc. is typically not included and I would argue might be a bit out of scope for the first steps for Wagtail in this GSoC and also this post. If there is strong interest happy to post more, but [this paper/article](https://publication2023.bits-und-baeume.org/#38
) gives a good starter on including embodied carbon and looking at the full life-cycle of software.

## 2. Possible optimization routes

To get the brainstorming process started I opted for putting out some typical optimization routes that we see and that maybe could also be applied for Wagtail in the GSoC:

- Swapping out architectures (Postgres Database for a SQLite database, NGINX for Apache, Memcache for Redis .... etc. )
- Using static code analyzers to make code level optimizations (Bundling SQL queries from loops to a single one, remove recursions / loops / nesting etc.)
  -  An interesting project that aims also for Python in an energy manner is [EcoCode](https://github.com/green-code-initiative)
- CPU level optimizations (Using vector instructions for bigger calculations)
- Swapping out libaries (Using numpy for bigger calculations, Using psycopg3 instead of psycopg2, using different image libraries, using / not using an ORM layer etc.)
- Optimizing frontend (Compressing images, reducing network transfer, lazy-loading etc.) 

However, also looking at the delivery side / auxillary tools might proof interesting
- Comparing testing tools like Ruff vs. Pylint
- Using a different Python runtime like cython
- Using a compiled version of Python (like [Mojo](https://www.modular.com/mojo) or [Codon](https://github.com/exaloop/codon))
- Reducing the cost of running CI/CD testing

## 3. Proper Benchmarking

The tools mentioned at the beginning we believe are very nice for prototyping and getting first results.

For instance to get the idea of the energy of a running Python script a one-liner with **perf_events** is usually enough to get a first glimpse.
[To demo the perf_events energy measurement principle here is an article on a related application from us](https://www.green-coding.berlin/case-studies/does-gzip-save-co2/)

In case of networking problems like image compression etc. typically formulas are used that operate on the transferred data. [CO2.js](https://developers.thegreenwebfoundation.org/co2js/overview/) from [The Green Web Foundation](https://www.thegreenwebfoundation.org/) is a splendid tool to do that.
To demo the principle here is another [article on getting energy metrics for network transfer](https://www.green-coding.berlin/case-studies/avif-vs-png/) from us.

For code runtime related energy issues, in which many of the aforementioned optimizations fall however, a proper benchmarking setup is needed that makes the appropriate amount of repetitions, configures the system correctly and also uses multiple means of measurement.
The python world has already good material on this, but I want to highlight the [article from pyperformance](https://pyperformance.readthedocs.io/usage.html#how-to-get-stable-benchmarks)

We at Green Coding Berlin have tried to bring most of these best practices and the best of available tooling into one bigger free and open-source project: [The Green Metrics Tool](https://www.green-coding.berlin/projects/green-metrics-tool/)

To best show how it works there is a demo where @thibaudcolas has setup [wagtail including a bakery demo already in a container setup](https://github.com/thibaudcolas/bakerydemo-gold-benchmark) which is then [measured with the Green Metrics Tool](https://metrics.green-coding.berlin/stats.html?id=0e3a289b-f083-4c69-b254-fad3130a03b9)

The tool can also compare measurements and make statistical evaluations. 
[Here an example](https://github.com/green-coding-berlin/django) where we are currently following the [Django](https://github.com/green-coding-berlin/django) project and their cost for their unit tests.
The data shows just repeated measurements for one commit atm, but as soon as more commits come in this will show the trend over time.

[Here an example](https://metrics.green-coding.berlin/compare.html?ids=97260fdc-01e8-44c3-b67c-417ff4e3d629,b7bb203a-d2c5-46a3-bf27-7d90d84da33e,3c9b123b-8b67-4bd5-a73d-1146e72d9856,fb6e3e27-6391-495b-8454-920fa922cfda,93c6411f-dd29-4d03-8774-2a31c7373ce9,b7086630-4b3e-454d-a339-7af4897d5a9c) for comparing Nextcloud with different databases 
![Screenshot 2023-06-10 at 10 03 53 AM](https://github.com/wagtail/wagtail/assets/250671/0578d885-3f45-49b3-b1d0-5d794b4061c9)
(SQLite vs. MariaDB ... green is SQLite)

The Green Metrics Tool can use many sensors like external power meters and is also installable locally ([See documentation](https://docs.green-coding.berlin/))

To make measurements more reliable and better comparable we offer [a free hosted instance](https://metrics.green-coding.berlin) at  where [a measurement can be handed in](https://metrics.green-coding.berlin/request.html).

The underlying architecture is a setup measurement cluster with an optimized Linux ([NOP Linux](https://github.com/green-coding-berlin/green-metrics-tool/issues/209))

For looking at changes in the CI/CD process it is typically not possible to make measurements directly, as the cloud vendors do not allow access to any sensors. Here we use a free and open source [machine learning model](https://www.green-coding.berlin/projects/cloud-energy/)

How this works in Github for instance can be seen on the [Eco-CI Page](https://www.green-coding.berlin/projects/eco-ci/)

## 4. Summary

I hope this article serves as a nice introduction into the topic of energy measurements for the GSoC and I am super stoked to see Wagtail tackling this topic :)

**Comment by ArneTR** [comment]
I added a link to the [Blue Angel Handbook](https://eco.kde.org/handbook/) earlier that I forgot to mention to link correctly :(

**Comment by mrchrisadams** [comment]
Thanks for this @ArneTR - this is really helpful guidance.

I have a question that might be a bit leftfield.

**Adapting Firefox Profiler to profile parts of python apps like Wagtail and others**

At FOSDEM, there was a talk by the folks who worked on the Firefox Profiler, where they explained how it can all kinds of .perf file provide detailed info, not just energy readings from the Firefox browsing sessions. In the browser, you can get per process, and within that, per thread (I think readings) and it goes into a lot of detail. you can see a recording below

https://fosdem.org/2023/schedule/event/energy_power_profiling_firefox/

Recently, there have been announcements about Python 3.12, having much richer support for profilers like perf - to the extent that you can see which python functions are being called inside the perf data now, rather than just the underlying system calls in a language like C.

These two blogs provide a bit more detail:

https://docs.python.org/3.12/howto/perf_profiling.html
https://www.petermcconnell.com/posts/perf_eng_with_py12/

I find it interesting because if you might use GMT ([green metric tool](https://github.com/green-coding-berlin/green-metrics-tool)) to understand hotspots to change at an architectural / _inter_-process level, I can see using something like these features for more detailed _intra_-process diagnosis.

We've already put some time into extending the Firefox profiler to make energy to carbon calculations using CO2.js, so that might also save time.

I can see this being useful for the following parts in the project outlined so far:
 
- [Sustainability considerations: hosting & infrastructure](https://github.com/orgs/wagtail/projects/30?pane=issue&itemId=30400369)
- [Sustainability considerations: backend / data processing](https://github.com/orgs/wagtail/projects/30?pane=issue&itemId=30400419)

I haven't looked into this in detail though yet - I'd welcome any pointers, or experiences from others here.
